WangscapeTest/TestCartesianPower.cpp

```diff
@@ -0,0 +1,262 @@
+{
+    auto it = cp.cbegin();
+    EXPECT_EQ(4, it.getValue(0));
+    EXPECT_EQ(4, it.getValue(1));
+    EXPECT_EQ(4, it.getValue(2));
+    EXPECT_EQ(4, it.getValue(3));
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerValueBegin1)
+{
+    auto it = cp.cbegin();
+    ++it;
+    EXPECT_EQ(5, it.getValue(0));
+    EXPECT_EQ(4, it.getValue(1));
+    EXPECT_EQ(4, it.getValue(2));
+    EXPECT_EQ(4, it.getValue(3));
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerValueBegin01)
+{
+    auto it = cp.cbegin();
+    std::advance(it, 3);
+    EXPECT_EQ(4, it.getValue(0));
+    EXPECT_EQ(5, it.getValue(1));
+    EXPECT_EQ(4, it.getValue(2));
+    EXPECT_EQ(4, it.getValue(3));
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerValueLast)
+{
+    auto it = cp.cbegin();
+    std::advance(it, 80);
+    EXPECT_EQ(6, it.getValue(0));
+    EXPECT_EQ(6, it.getValue(1));
+    EXPECT_EQ(6, it.getValue(2));
+    EXPECT_EQ(6, it.getValue(3));
+    ++it;
+    EXPECT_EQ(it, cp.cend());
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerCoordinates)
+{
+    auto it = cp.cbegin();
+    std::advance(it, 15);
+
+    EXPECT_EQ(0, it.getCoordinate(0));
+    EXPECT_EQ(2, it.getCoordinate(1));
+    EXPECT_EQ(1, it.getCoordinate(2));
+    EXPECT_EQ(0, it.getCoordinate(3));
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerArrow)
+{
+    auto it = cp.cbegin();
+    EXPECT_EQ(cp.power, it->size());
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerValues)
+{
+    auto it = cp.cbegin();
+    std::advance(it, 15);
+    const std::vector<int>& values = *it;
+
+    EXPECT_EQ(4, values[0]);
+    EXPECT_EQ(6, values[1]);
+    EXPECT_EQ(5, values[2]);
+    EXPECT_EQ(4, values[3]);
+
+    for (size_t i = 0; i < values.size(); i++)
+    {
+        EXPECT_EQ(values[i], it.getValue(i));
+    }
+
+    for (size_t i = 0; i < values.size(); i++)
+    {
+        EXPECT_EQ(values[i], *it.getIterator(i));
+    }
+
+    for (size_t i = 0; i < values.size(); i++)
+    {
+        EXPECT_EQ(values[i], *it.getIterators()[i]);
+    }
+
+}
+TEST_F(TestCartesianPower, TestCoordinatePackerDomain)
+{
+    tilegen::CoordinatePacker<size_t> coord_packer0(0);
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(-1));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(0));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(1));
+
+    tilegen::CoordinatePacker<size_t> coord_packer1(1);
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(-1));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(1));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(2));
+
+    tilegen::CoordinatePacker<size_t> coord_packer2(2);
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(-1));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(2));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(3));
+
+    tilegen::CoordinatePacker<size_t> coord_packer10(10);
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(-1));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(10));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(11));
```

Comment:
Copypaste fail. This should be replaced with a loop that checks the correct object in each case.